src/zkp/pienc.rs

```diff
@@ -27,196 +41,269 @@ use merlin::Transcript;
-            let rhs = self.C.modmul(
-                &modpow(&self.S, &e, &input.setup_params.N),
+            let rhs = self.plaintext_mask_commit.modmul(
+                &modpow(&self.plaintext_commit, &e, &input.setup_params.N),
                 &input.setup_params.N,
             );
             lhs == rhs
         };
-        if !eq_check_2 {
-            return verify_err!("eq_check_2 failed");
+        if !responses_match_commitments {
+            return verify_err!("response validation check (second quality check) failed");
         }
 
-        // Do range check
+        // Make sure the ciphertext response is in range
         let bound = BigNumber::one() << (ELL + EPSILON);
-        if self.z1 < -bound.clone() || self.z1 > bound {
-            return verify_err!("self.z1 > bound check failed");
+        if self.plaintext_response < -bound.clone() || self.plaintext_response > bound {
+            return verify_err!("bounds check on plaintext response failed");
         }
 
         Ok(())
     }
 }
 
+impl PiEncProof {
+    /// Update the [`Transcript`] with all the commitment values used in the proof.
+    fn fill_out_transcript(
+        transcript: &mut Transcript,
+        input: &PiEncInput,
+        plaintext_mask: &BigNumber,
+        plaintext_commit: &BigNumber,
+        ciphertext_mask: &Ciphertext,
+        plaintext_mask_commit: &BigNumber,
+    ) -> Result<()> {
+        transcript.append_message(b"PiEnc CommonInput", &serialize!(&input)?);
+        transcript.append_message(b"alpha", &plaintext_mask.to_bytes());
+        transcript.append_message(
+            b"(S, A, C)",
+            &[
+                plaintext_commit.to_bytes(),
+                ciphertext_mask.to_bytes(),
+                plaintext_mask_commit.to_bytes(),
+            ]
+            .concat(),
+        );
+        Ok(())
+    }
+}
+
 #[cfg(test)]
 mod tests {
     use super::*;
-    use crate::{paillier::PaillierDecryptionKey, utils::random_plusminus_by_size_with_minimum};
+    use crate::{paillier::DecryptionKey, utils::random_plusminus_by_size_with_minimum};
 
     fn random_paillier_encryption_in_range_proof<R: RngCore + CryptoRng>(
         rng: &mut R,
         k: &BigNumber,
     ) -> Result<()> {
-        let (decryption_key, _, _) = PaillierDecryptionKey::new(rng)?;
+        let (decryption_key, _, _) = DecryptionKey::new(rng)?;
         let pk = decryption_key.encryption_key();
 
         let (K, rho) = pk.encrypt(rng, k)?;
         let setup_params = ZkSetupParameters::gen(rng)?;
 
         let input = PiEncInput {
             setup_params,
-            pk,
-            K,
+            encryption_key: pk,
+            ciphertext: K,
         };
 
-        let proof = PiEncProof::prove(rng, &input, &PiEncSecret { k: k.clone(), rho })?;
+        let proof = PiEncProof::prove(
+            rng,
+            &input,
+            &PiEncSecret {
+                plaintext: k.clone(),
```

Comment:
Very minor, since it's in a test, but is a clone needed here?

Reply:
Good call, it's not.